src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +30,155 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn sleep(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Debug)]
+struct AwaitingCaller {
+    wake_up_time: DateTime<Utc>,
+    waker_tx: Option<oneshot::Sender<()>>,
+}
+
+impl Eq for AwaitingCaller {}
+
+impl PartialEq<Self> for AwaitingCaller {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl PartialOrd<Self> for AwaitingCaller {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for AwaitingCaller {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.wake_up_time.cmp(&other.wake_up_time)
+    }
+}
+
+#[derive(Debug)]
+struct SystemTimeSourceStubState {
+    t: DateTime<Utc>,
+    awaiting_callers: BinaryHeap<Reverse<AwaitingCaller>>,
+}
+
 #[derive(Debug, Clone)]
-pub struct SystemTimeSourceStub {
-    t: Arc<Mutex<Option<DateTime<Utc>>>>,
+pub struct FakeSystemTimeSource {
+    state: Arc<Mutex<SystemTimeSourceStubState>>,
 }
 
-impl SystemTimeSourceStub {
+impl FakeSystemTimeSource {
     pub fn new() -> Self {
+        let state = SystemTimeSourceStubState {
+            t: Utc::now(),
+            awaiting_callers: BinaryHeap::new(),
+        };
+
         Self {
-            t: Arc::new(Mutex::new(None)),
+            state: Arc::new(Mutex::new(state)),
         }
     }
 
     pub fn new_set(t: DateTime<Utc>) -> Self {
+        let state = SystemTimeSourceStubState {
+            t,
+            awaiting_callers: BinaryHeap::new(),
+        };
+
         Self {
-            t: Arc::new(Mutex::new(Some(t))),
+            state: Arc::new(Mutex::new(state)),
         }
     }
 
     pub fn set(&self, t: DateTime<Utc>) {
-        *self.t.lock().unwrap() = Some(t);
+        let mut state = self.state.lock().unwrap();
+
+        if state.t > t {
+            panic!(
+                "The previous time [{}] is more than new time [{t}]",
```

Comment:
"New time is earlier than previously seen".

Although, if you split the stub and fake, as I asked, I don't think we need a `set` public method at all, and could do everything in `advance` implementation.

Reply:
Fixed